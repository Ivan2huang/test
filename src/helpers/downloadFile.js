import { fetchFile } from './fetch';

const SUCCESS = 'SUCCESS';
const ERROR = 'ERROR';

const downloadFileAsync = async (url, fileName) => {
  const response = await fetchFile(url);
  if (response) {
    const blob = new Blob([response]);
    if (window.navigator.msSaveOrOpenBlob) {
      window.navigator.msSaveOrOpenBlob(blob, fileName);
      return SUCCESS;
    }
    const objectUrl = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = objectUrl;
    link.setAttribute('download', fileName);
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    return SUCCESS;
  }
  return ERROR;
};

export default downloadFileAsync;
