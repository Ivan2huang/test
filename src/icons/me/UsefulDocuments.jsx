import SvgIcon from '@material-ui/core/SvgIcon';
import React from 'react';
import * as PropTypes from 'prop-types';
import colors from '../../themes/colors';
import { ICON_SIZE } from '../../themes/theme';

const UsefulDocuments = ({ width, height, ...rest }) => (
  <SvgIcon style={{ width, height }} {...rest}>
    <path
      d="M17.5997 0H1.33301V24H22.6663V5.06667L17.5997 0ZM20.4037 5.06667H17.5997V2.26267L20.4037 5.06667ZM21.0663 22.4H2.93301V1.6H15.9997V6.66667H21.0663V22.4Z"
      fill={colors.icon || colors.mediumEmphasis}
    />
  </SvgIcon>
);

UsefulDocuments.propTypes = {
  width: PropTypes.string,
  height: PropTypes.string,
};

UsefulDocuments.defaultProps = {
  width: ICON_SIZE,
  height: ICON_SIZE,
};

export default UsefulDocuments;
