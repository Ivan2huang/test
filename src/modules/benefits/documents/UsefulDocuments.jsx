import React from 'react';
import { injectIntl } from 'react-intl';
import { Box } from '@material-ui/core';
import PropTypes from 'prop-types';
import Typography from '../../../uiComponents/Typography';
import FileItems from '../../../uiComponents/FileItems';
import Grid from '../../../uiComponents/Grid';
import GridItem from '../../../uiComponents/GridItem';

import { logAction } from '../../../helpers/firebase';
import { CATEGORIES } from '../../../constants/analytics';
import downloadFile from '../../../helpers/downloadFile';

const UsefulDocuments = ({ intl, usefulDocuments }) => (
  <>
    <Box>
      <Typography type="style-2">
        {intl.formatMessage({
          id: 'me.tabs.usefulDocuments.title',
          defaultMessage: 'Forms and Documents',
        })}
      </Typography>
    </Box>
    <Grid>
      <GridItem columns={{ xs: 12, md: 8 }}>
        <Box mt={{ xs: -2, md: 0 }}>
          {usefulDocuments.length > 0 && (
            <FileItems
              files={usefulDocuments}
              onClick={file => {
                logAction({
                  category: CATEGORIES.DOCUMENTS_PAGE,
                  action: 'View useful document',
                  document_name: file.name,
                });
                downloadFile(file.url, `${file.name}.${file.contentType}`);
              }}
            />
          )}
        </Box>
      </GridItem>
    </Grid>
  </>
);

UsefulDocuments.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  usefulDocuments: PropTypes.arrayOf(
    PropTypes.exact({
      name: PropTypes.string.isRequired,
      url: PropTypes.string.isRequired,
      contentType: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default injectIntl(UsefulDocuments);
