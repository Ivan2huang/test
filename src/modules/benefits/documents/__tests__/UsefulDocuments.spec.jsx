/* eslint-disable react/prop-types */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import UsefulDocuments from '../UsefulDocuments';
import withTheme from '../../../../themes/withThemeProvider';
import withIntl from '../../../../i18n/withIntlProvider';
import downloadFile from '../../../../helpers/downloadFile';

jest.mock(
  '../../../../uiComponents/Typography',
  () => ({ children, props }) => (
    <div {...props}>
      Typography Component
      {children}
    </div>
  ),
);

jest.mock('../../../../uiComponents/FileItems', () => ({ files, onClick }) => (
  <div>
    File Items
    {files.map((file, index) => (
      <button
        type="button"
        key={file.name}
        data-testid={`file-${index}`}
        onClick={() => onClick(file)}
      >
        {file.name}
      </button>
    ))}
  </div>
));

jest.mock(
  '../../../../uiComponents/GridItem',
  () => ({ columns, children }) => (
    <div data-xs={columns.xs} data-sm={columns.sm} data-md={columns.md}>
      <span>GridItem Component</span>
      <div>{children}</div>
    </div>
  ),
);
jest.mock('../../../../uiComponents/Grid', () => ({ children }) => (
  <div>
    <span>Grid Component</span>
    <div>{children}</div>
  </div>
));

jest.mock('../../../../helpers/helpers', () => ({
  formatMessage: jest.fn((intl, id, defaultMessage) => defaultMessage),
}));

jest.mock('../../../../helpers/downloadFile');

afterEach(() => {
  jest.clearAllMocks();
});

describe('useful documents component', () => {
  const props = {
    usefulDocuments: [{ name: 'name', url: 'url', contentType: 'pdf' }],
  };

  const Component = withTheme(withIntl(UsefulDocuments));
  const setup = (newProps = { ...props }) => {
    return render(<Component {...newProps} />);
  };

  it('should match snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot', () => {
    const { getByTestId } = setup();
    const fileButton = getByTestId('file-0');

    fireEvent.click(fileButton);

    expect(downloadFile).toHaveBeenCalledTimes(1);
    expect(downloadFile).toHaveBeenCalledWith('url', 'name.pdf');
  });
});
