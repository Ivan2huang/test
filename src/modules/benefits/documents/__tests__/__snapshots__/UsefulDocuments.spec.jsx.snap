// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useful documents component should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div>
      Typography Component
      Forms and Documents
    </div>
  </div>
  <div>
    <span>
      Grid Component
    </span>
    <div>
      <div
        data-md="8"
        data-xs="12"
      >
        <span>
          GridItem Component
        </span>
        <div>
          <div
            class="MuiBox-root MuiBox-root-3"
          >
            <div>
              File Items
              <button
                data-testid="file-0"
                type="button"
              >
                name
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
